EIPS/eip-2537.md

```diff
@@ -0,0 +1,301 @@
+Motivation section covers a total motivation to have operations over BLS12-381 curve available. We also extend a rationale for move specific fine points.
+
+#### Multiexponentiation as a separate call
+
+Explicit separate multiexponentiation operation that allows one to save execution time (so gas) by both the algorithm used (namely Peppinger algorithm) and (usually forgotten) by the fact that `CALL` operation in Ethereum is expensive (at the time of writing), so one would have to pay non-negigible overhead if e.g. for multiexponentiation of `100` points would have to call the multipication precompile `100` times and addition for `99` times (roughly `138600` would be saved).
+
+#### Explicit subgroup checks
+
+Subgroup checks are made optional both due to the fact that they can be performed explicitly by the caller (by multiplication operation) and due to the fact that subgroup checks in G2 that are expensive are not required in most of the cases cause G2 points are usually "hardcoded" and not supplied by the untrusted third party.
+
+Concretely, G2 subgroup check is around `55000` gas.
+
+## Backwards Compatibility
+<!--All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The EIP must explain how the author proposes to deal with these incompatibilities. EIP submissions without a sufficient backwards compatibility treatise may be rejected outright.-->
+There are no backward compatibility questions.
+
+## Test Cases
+<!--Test cases for an implementation are mandatory for EIPs that are affecting consensus changes. Other EIPs can choose to include links to test cases if applicable.-->
+
+Due to the large test parameters space we first provide properties that various operations must satisfy. We use additive notation for point operations, capital letters (`P`, `Q`) for points, small letters (`a`, `b`) for scalars. Generator for G1 is labeled as `G`, generator for G2 is labeled as `H`, otherwise we assume random point on a curve in a correct subgroup. `0` means either scalar zero or point of infinity. `1` means either scalar one or multiplicative identity. `group_order` is a main subgroup order. `e(P, Q)` means pairing operation where `P` is in G1, `Q` is in G2. 
+
+Requeired properties for basic ops (add/multiply):
+
+- Commutativity: `P + Q = Q + P`
+- Additive negation: `P + (-P) = 0`
+- Doubling `P + P = 2*P`
+- Subgroup check: `group_order * P = 0`
+- Trivial multiplication check: `1 * P = P`
+- Multiplication by zero: `0 * P = 0`
+- Multiplication by the unnormalized scalar `(scalar + group_order) * P = scalar * P`
+
+Required properties for pairing operation:
+- Degeneracy `e(P, 0*Q) = e(0*P, Q) = 1` 
+- Bilinearity `e(a*P, b*Q) = e(a*b*P, Q) = e(P, a*b*Q)` (internal test, not visible through ABI)
+
+Test vector for all operations are expanded in this [gist](https://gist.github.com/shamatar/f82106e4b3a957761c439eb6024e5300) until it's final.
```

Comment:
Gist only contains one vector now -- please add some more (a few hundred maybe ?)

Reply:
Due to breaking ABI change I'll have to purge the gist now and redo all the work, but yes, it'll be added